go-controller/hybrid-overlay/pkg/controller/master.go

```
@@ -0,0 +1,203 @@
+package controller
+
+import (
+	"bytes"
+	"fmt"
+	"net"
+
+	"github.com/sirupsen/logrus"
+
+	"github.com/ovn-org/ovn-kubernetes/go-controller/hybrid-overlay/pkg/types"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/hybrid-overlay/pkg/util"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/config"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/factory"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/kube"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovn"
+
+	"github.com/openshift/origin/pkg/util/netutils"
+	kapi "k8s.io/api/core/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+// MasterController is the master hybrid overlay controller
+type MasterController struct {
+	kube      *kube.Kube
+	allocator []*netutils.SubnetAllocator
+}
+
+// NewMaster a new master controller that listens for node events
+func NewMaster(clientset kubernetes.Interface, subnets []config.CIDRNetworkEntry) (*MasterController, error) {
+	m := &MasterController{
+		kube: &kube.Kube{KClient: clientset},
+	}
+
+	alreadyAllocated := make([]string, 0)
+	existingNodes, err := m.kube.GetNodes()
```

Correct me if I'm wrong, but will GetNodes have its cache sunk at this time? Do we need to worry about partial results here?

---

(We could move a decent amount of this logic to the `Sync()` function if not.

---

@squeed theoretically, but Sync() cannot return an error and kill the process unless we want to panic(). The reason it's still in the startup part is to ensure we can return an error when setting up our allocator.

But also, yes the cache is sunk. The Factory ensures that caches are synced before it returns from NewWatchFactor() in ovnkube.go.